### PR TITLE
Dimension API 1.16.2

### PR DESCRIPTION
**Pull request by shartte**
This makes the dimension API work again.

Vanilla no longer creates portals for dimensions other than the nether, so the entire Placement Handling is not necessarily needed for mods that just want to port to a specific point in their dimension.

I also introduced two different term here:
- ~~Untargeted Teleportation means teleporting to a dimension without specifying a destination point (nether portals, end portals, Vanilla `moveToWorld` API)~~
- Targeted Teleportation means teleporting to a dimension while specifying the destination point in the target dimension fully (this will apply to most mods porting players into their own dimensions using teleports or various other methods)

~~A modded dimension needs to opt-in to support the untargeted teleportation case, by specifying a handler for it. That handler would then carry out tasks such as:~~
- ~~Determining the actual destination, for example by deriving it from the entities current position~~
- ~~Placing a portal - if desired~~

~~To support special use cases such as performing an untargeted teleportation into a mod dimension _WITHOUT_ placing a portal, the API supports optionally passing attributes along with the teleportation request, which the handler can then inspect.~~

**edit**: I decided to rip out the untargeted teleportation for now since it seems like a niche use case and would require us to replicate Vanilla's coordinate mapping/scaling behavior to teleport to the overworld in an untargetted fashion.

Additionally, I introduced two new methods to allow mods to mark their dimension and dimension_type JSON files as stable so that they do not trigger the experimental warning screen anymore. This requires any used Chunk Generators to use a fully stable codec (the test mod's Codec was experimental, which cost me quite a bit of time).

The mixin into RegistryOps could potentially be replaced with dynamic registry events if those are merged, but is currently independent.

### CODE REVIEW

**Comment by shartte** [comment]
This API might be affected by https://bugs.mojang.com/browse/MC-195468 for dimension options defined in JSON.

https://discord.com/channels/507304429255393322/608088354042544139/747657399271293082

**Comment by shartte** [comment]
I wrote a Mixin that works around the mentioned bug, and will also fix this for biomes and other worldgen settings.

p.s.: It does NOT fix referencing datapack biomes in server.properties if that matters.

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/ServerBugfixMixin.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ * dimension testmod, and no world directory. If the dimension is available (i.e. in /execute in, or via
+ * the testmod's commands), then the bug is fixed and this Mixin can be removed.
+ */
+@Mixin(value = Main.class, remap = false)
+public class ServerBugfixMixin {
+	private static LevelStorage.Session fabric_session;
+
+	private static DynamicRegistryManager.Impl fabric_dynamicRegistry;
+
+	private static RegistryOps<Tag> fabric_registryOps;
```
Comment (TheGlitch76) [inline fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/ServerBugfixMixin.java:58]: tiny nitpick but shouldn't these be `@Unique`?

Comment (shartte) [reply fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/ServerBugfixMixin.java:58]: I don't know? Maybe? :-D
Doesn't @Unique just scramble the name and make the fabric_ prefix superflous? I thought adding the fabric_ prefix as per mod-id prefixing conventions was already sufficient.

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/RegistryOpsMixin.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ * This convoluted Mixins purpose is to mark only <em>specific</em> but not <em>all</em> dimension types that are
+ * loaded via the Registry Codec to be stable. This is required to hide the experimental warning screen when
+ * a world is joined.
+ *
+ * <p>To do this in the most compatible way possible, we capture the identifier of the element that is being loaded
+ * in a thread-local, and then modify the lifecycle of the data result based on that identifier.
+ */
+@Mixin(RegistryOps.class)
+public class RegistryOpsMixin {
+	private ThreadLocal<RegistryKey<?>> fabric_currentlyLoading;
```
Comment (liach) [inline fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/RegistryOpsMixin.java:51]: :+1:

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/ServerBugfixMixin.java (hunk 1 above)
Comment (liach) [reply fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/ServerBugfixMixin.java:58]: Agree with shartte. unique annotation is overrated.

Comment (TheBrokenRail) [reply fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/ServerBugfixMixin.java:58]: Why not just use @Unique and ignore the prefix?

**Comment by qouteall** [comment]
Currently, there is still something that Fabric Dimension API needs
1. To avoid mod dimensions to mess with DFU
Datapack-added dimensions will be stored in level.dat, including mod dimensions. If the dimension mod is removed, DFU cannot recognize it and throw and error. Then the nether dimension and end dimension will be swallowed.
For example, I tested bumblezone, create a new world, save, remove bumblezone, open the world, then the nether and the end vanished.
One way to solve it is to remove the mod dimensions from GeneratorOptions before saving to level.dat
2. Fix the issue that the seed should be hardcoded
This can be solved by providing a way to add dimensions at runtime (not by datapack). This makes it possible for a mod to disable a dimension according to config.

**Comment by liach** [comment]
> To avoid mod dimensions to mess with DFU

I personally suggest patching codec to handle these mod dimensions with special logic.

**Comment by shartte** [comment]
Before investing more time in this, I'd really like reviews by @modmuss50 or @sfPlayer1 to see where this should go.

p.s.: I personally would prefer just dropping API-support for Mods adding dimensions via JSON and move over to fully code-based. This would also expose properties currently not exposed to JSON (i.e. sky properties), but we'd need to write a builder for DimensionType.

**Comment by shartte** [comment]
To document clearly what is happening:

- Run MinecraftServer with dimension-testmod as the module
- Let the world generate fully, run save-all and exit server
- Run MinecraftServer with dimension-api as the module (so that the test-mod dimension is gone)
- You'll see the following error:
```
[17:27:19] [main/ERROR] (Minecraft) WorldGenSettings: Unknown registry key: fabric_dimension:void; Failed to parse fabric_dimension:dimension_type/void_type.json file: fabric_dimension:dimension_type/void_type.json missed input: {"fabric_dimension:void":{generator:{biome_source:{biome:"minecraft:plains",type:"minecraft:fixed"},custom_bool:1b,type:"fabric_dimension:void"},type:"fabric_dimension:void_type"}}
```
- ~~When the server is started, run `execute in minecraft:the_nether run list`, and you'll get `Unknown dimension 'minecraft:the_nether'`~~ This is actually not true. Running this command will still print the player list.

After looking at it again locally: Vanilla seems to drop the dimensions after the first one with an error, because that's what the "partial result" of parsing the registry is.

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/ServerBugfixMixin.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import net.minecraft.util.dynamic.RegistryOps;
+import net.minecraft.util.registry.DynamicRegistryManager;
+import net.minecraft.world.gen.GeneratorOptions;
+import net.minecraft.world.level.LevelInfo;
+import net.minecraft.world.level.LevelProperties;
+import net.minecraft.world.level.storage.LevelStorage;
+
+/**
+ * This Mixin aims to solve a Minecraft Vanilla bug where datapacks are ignored during creation of the
+ * initial LevelProperties when a dedicated server creates a completely new level.
```
Comment (modmuss50) [inline fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/ServerBugfixMixin.java:37]: Why does this specifically need a fix here? The javadoc says you fixed a bug but doesnt say why you fixed it?

File: fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/RegistryOpsMixin.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import net.minecraft.util.dynamic.RegistryOps;
+import net.minecraft.util.registry.MutableRegistry;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.util.registry.RegistryKey;
+import net.minecraft.world.dimension.DimensionType;
+
+import net.fabricmc.fabric.impl.dimension.FabricDimensionInternals;
+
+/**
+ * This convoluted Mixins purpose is to mark only <em>specific</em> but not <em>all</em> dimension types that are
```
Comment (modmuss50) [inline fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/mixin/dimension/RegistryOpsMixin.java:42]: Do we really want to be able to mark dims as stable? We decided in the past to not do this.

**Comment by oliviathevampire** [comment]
Can this be updated to either 1.16.4 or 20w45a?

**Comment by modmuss50** [comment]
This was on my list of things to update and get in. Might as well keep the PR open.

**Comment by shartte** [comment]
This has been going stale for over two months. I have no plans to work on this any further. Please open a new PR if desired.

**Pull request closed without merging**